File: Hard-Array/PascalTrinagle.java
// WAP to print Pascal's Triangle
import java.util.ArrayList;
import java.util.Scanner;

class PascalTrinagle
{
    public static ArrayList<Integer> nthRowOfPascalTriangle(int n)
    {
        // Adjust to 0-indexed internally
        int rowIndex = n - 1; 
        ArrayList<Integer> row = new ArrayList<>();
        int number = 1;
        for (int k = 0; k <= rowIndex; k++)
        {
            row.add(number);
            number = number * (rowIndex - k) / (k + 1);
        }
        return row;
    }
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of rows for Pascal's Triangle: ");
        int n = sc.nextInt();
        
        for (int i = 0; i < n; i++) {
            int number = 1; // First element in each row is always 1
            System.out.print(" ".repeat(n - i - 1)); // Print leading spaces
            
            for (int j = 0; j <= i; j++) {
                System.out.print(number + " ");
                number = number * (i - j) / (j + 1); // Calculate next number in the row
            }
            System.out.println(); // Move to the next line after each row
        }
    }
}

File: Hard-Array/PascalTrinagle.py
# WAP to print the pascal triangle

# Pascal's Triangle is a triangular array of numbers where each number is the sum of the two numbers directly above it in the previous row. The first and last number of each row is always 1. The nth row of Pascal's Triangle contains the binomial coefficients, which can be calculated using the formula: C(n, k) = n! / (k! * (n - k)!), where n is the row number and k is the position in the row.
def pascal_triangle(nums):
    # Loop through each row of the triangle
    for i in range(nums):
        # Print spaces to center the triangle
        for j in range(nums - i - 1):
            print(" ", end="")
        
        # Loop through each element in the row and calculate the binomial coefficient
        for j in range(i + 1):
            print(f"{factorial(i) // (factorial(j) * factorial(i - j))} ", end=" ")
        print()

# A helper function to calculate the factorial of a number
def factorial(num):
    # Base case: factorial of 0 or 1 is 1
    if num == 0 or num == 1:
        return 1
    # Recursive case: factorial of n is n multiplied by factorial of (n - 1)
    else:
        return num * factorial(num - 1)

num = int(input("Enter the number of rows for Pascal's Triangle: "))
pascal_triangle(num)
print("Pascal's Triangle printed successfully!")
